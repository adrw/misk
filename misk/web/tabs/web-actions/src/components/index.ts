export * from "./CommonComponents"
export * from "./FilterWebActionsComponent"
export * from "./RequestFormComponent"
export * from "./SendRequestCollapseComponent"
export * from "./WebActionsComponent"
