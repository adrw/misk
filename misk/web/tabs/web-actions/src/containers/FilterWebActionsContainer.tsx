import { ControlGroup, HTMLSelect, InputGroup } from "@blueprintjs/core"
import { onChangeFnCall } from "@misk/simpleredux"
import * as React from "react"
import { connect } from "react-redux"
import {
  IDispatchProps,
  mapDispatchToProps,
  mapStateToProps,
  WebActionInternalLabel
} from "../ducks"

const FilterWebActionsContainer = (props: { tag: string } & IDispatchProps) => {
  const FilterSelectOptions = [...Object.keys(WebActionInternalLabel)]
  const filterTag = `${props.tag}::Filter`
  return (
    <ControlGroup fill={true}>
      <HTMLSelect
        large={true}
        onChange={onChangeFnCall(
          props.simpleFormInput,
          `${filterTag}::HTMLSelect`
        )}
        options={FilterSelectOptions}
      />
      <InputGroup
        large={true}
        onChange={onChangeFnCall(props.simpleFormInput, `${filterTag}::Input`)}
        placeholder={"Filter Web Actions"}
      />
      {/* TODO(adrw) Use multiselect to do autocomplete filters */}
    </ControlGroup>
  )
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(FilterWebActionsContainer)
