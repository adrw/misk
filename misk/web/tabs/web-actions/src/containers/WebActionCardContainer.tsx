import { Card, H3, H5, MenuItem } from "@blueprintjs/core"
import { FlexContainer, WrapTextContainer } from "@misk/core"
import * as React from "react"
import { connect } from "react-redux"
import {
  CodeTag,
  Column,
  FloatLeft,
  Header,
  Metadata,
  MetadataCollapse,
  MetadataMenu,
  MethodTag,
  RequestResponeContentTypes,
  SendRequestCollapseComponent
} from "../components"
import {
  IDispatchProps,
  IState,
  IWebActionInternal,
  mapDispatchToProps,
  mapStateToProps
} from "../ducks"

/**
 * Web Action Card rendered for each bound Web Action
 */
const WebActionCardContainer = (
  props: { action: IWebActionInternal; tag: string } & IState & IDispatchProps
) => (
  <div>
    <Card>
      <Header>
        {props.action.dispatchMechanism.map(m => (
          <MethodTag key={m} method={m} />
        ))}
        <FloatLeft>
          <H3>{props.action.name}</H3>
        </FloatLeft>
        <FloatLeft>
          <CodeTag large={true}>{props.action.pathPattern}</CodeTag>
        </FloatLeft>
      </Header>
      {props.action.nonAccessOrTypeFunctionAnnotations.map((a, index) => (
        <H5 key={index}>{a}</H5>
      ))}
      <FlexContainer>
        <Column>
          <MetadataMenu>
            <Metadata
              content={props.action.function}
              label={"Function"}
              tooltip={props.action.function}
            />
            <Metadata
              content={props.action.allowedServices}
              label={"Services"}
              tooltip={props.action.allowedServices}
            />
            <Metadata
              content={props.action.allowedRoles}
              label={"Roles"}
              tooltip={props.action.allowedRoles}
            />
            <Metadata
              content={props.action.authFunctionAnnotations[0]}
              label={"Access"}
              tooltip={props.action.authFunctionAnnotations[0]}
            />
          </MetadataMenu>
        </Column>
        <Column>
          <MetadataMenu>
            <Metadata
              content={<RequestResponeContentTypes action={props.action} />}
              label={"Content Types"}
              tooltip={<RequestResponeContentTypes action={props.action} />}
            />
            <MetadataCollapse
              {...props}
              content={"Application Interceptors"}
              label={`(${props.action.applicationInterceptors.length})`}
              tag={`${props.tag}::ApplicationInterceptors`}
            >
              {props.action.applicationInterceptors.map((ai, index) => (
                <MenuItem
                  key={index}
                  text={<WrapTextContainer>{ai}</WrapTextContainer>}
                />
              ))}
            </MetadataCollapse>
            <MetadataCollapse
              {...props}
              content={"Network Interceptors"}
              label={`(${props.action.networkInterceptors.length})`}
              tag={`${props.tag}::NetworkInterceptors`}
            >
              {props.action.networkInterceptors.map((ni, index) => (
                <MenuItem
                  key={index}
                  text={<WrapTextContainer>{ni}</WrapTextContainer>}
                />
              ))}
            </MetadataCollapse>
            <MetadataCollapse
              {...props}
              content={"Send a Request"}
              label={""}
              tag={`${props.tag}::ButtonSendRequest`}
            >
              <span />
            </MetadataCollapse>
          </MetadataMenu>
        </Column>
      </FlexContainer>
      <SendRequestCollapseComponent {...props} />
    </Card>
    <br />
  </div>
)

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(WebActionCardContainer)
