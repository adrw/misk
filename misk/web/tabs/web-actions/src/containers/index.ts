export { default as TabContainer } from "./TabContainer"
export { default as WebActionsContainer } from "./WebActionsContainer"
