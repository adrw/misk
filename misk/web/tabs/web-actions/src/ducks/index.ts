import {
  dispatchSimpleForm,
  dispatchSimpleNetwork,
  IDispatchSimpleForm,
  IDispatchSimpleNetwork,
  ISimpleFormImmutableState,
  ISimpleFormState,
  ISimpleNetworkImmutableState,
  ISimpleNetworkState,
  SimpleFormReducer,
  SimpleNetworkReducer,
  simpleRootSelector,
  watchSimpleFormSagas,
  watchSimpleNetworkSagas
} from "@misk/simpleredux"
import {
  connectRouter,
  LocationChangeAction,
  RouterState
} from "connected-react-router"
import { History } from "history"
import { AnyAction, combineReducers, Reducer } from "redux"
import { all, AllEffect, fork } from "redux-saga/effects"
import {
  dispatchWebActions,
  IDispatchWebActions,
  IWebActionsState,
  watchWebActionsSagas,
  WebActionsReducer
} from "./webActions"
export * from "./webActions"

/**
 * Redux Store State
 */
export interface IState {
  webActions: IWebActionsState
  router: Reducer<RouterState, LocationChangeAction>
  simpleForm: ISimpleFormState
  simpleNetwork: ISimpleNetworkState
}

/**
 * Dispatcher
 */
export interface IDispatchProps
  extends IDispatchSimpleForm,
    IDispatchSimpleNetwork,
    IDispatchWebActions {}

export const rootDispatcher: IDispatchProps = {
  ...dispatchSimpleForm,
  ...dispatchSimpleNetwork,
  ...dispatchWebActions
}

/**
 * State Selectors
 */
export const rootSelectors = (state: IState) => ({
  simpleForm: simpleRootSelector<IState, ISimpleFormImmutableState>(
    "simpleForm",
    state
  ),
  simpleNetwork: simpleRootSelector<IState, ISimpleNetworkImmutableState>(
    "simpleNetwork",
    state
  ),
  webActions: state.webActions
})

/**
 * Reducers
 */
export const rootReducer = (history: History): Reducer<any, AnyAction> =>
  combineReducers({
    router: connectRouter(history),
    simpleForm: SimpleFormReducer,
    simpleNetwork: SimpleNetworkReducer,
    webActions: WebActionsReducer
  })

/**
 * Sagas
 */
export function* rootSaga(): IterableIterator<AllEffect> {
  yield all([
    fork(watchWebActionsSagas),
    fork(watchSimpleFormSagas),
    fork(watchSimpleNetworkSagas)
  ])
}
