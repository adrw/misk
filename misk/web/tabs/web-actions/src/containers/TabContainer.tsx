import { H1 } from "@blueprintjs/core"
import * as React from "react"
import { connect } from "react-redux"
import { WebActionsComponent } from "../components"
import { IDispatchProps, IState, rootDispatcher, rootSelectors } from "../ducks"

class TabContainer extends React.Component<IState & IDispatchProps, IState> {
  componentDidMount() {
    this.props.webActionsMetadata()
  }

  render() {
    return (
      <div>
        <H1>Web Actions</H1>
        <WebActionsComponent tag={"WebActions"} {...this.props} />
      </div>
    )
  }
}

const mapStateToProps = (state: IState) => rootSelectors(state)

const mapDispatchToProps = {
  ...rootDispatcher
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TabContainer)
