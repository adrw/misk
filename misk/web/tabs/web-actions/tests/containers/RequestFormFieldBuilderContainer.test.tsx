import HTTPMethod from "http-method-enum"
import React from "react"
import { cleanup } from "react-testing-library"
import { RequestFormFieldBuilderContainer } from "../../src/containers"
import { generateTypesMetadata } from "../../src/ducks"
import {
  nonTypedActionAPI,
  nonTypedActionInternal,
  testTypes
} from "../testUtilities"
import { renderWithRedux } from "../upstreamableTestUtilities"

jest.mock("@misk/core", () => {
  const miskCore = require.requireActual("@misk/core")
  const mock = {
    ...miskCore,
    Column: jest.fn((props: { children: any }) => (
      <div className={"Column"}>{props.children}</div>
    )),
    FlexContainer: jest.fn((props: { children: any }) => (
      <div className={"FlexContainer"}>{props.children}</div>
    ))
  }
  return mock
})

describe("RequestFormFieldBuilderContainer", () => {
  afterEach(cleanup)
  it("Doesn't fail on empty metadata", () => {
    const typesMetadata = generateTypesMetadata(nonTypedActionAPI)
    const { asFragment } = renderWithRedux(
      <RequestFormFieldBuilderContainer
        action={nonTypedActionInternal}
        id={"0"}
        tag={"WebActions"}
        typesMetadata={typesMetadata}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("Renders single body input", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST
    })
    const { asFragment } = renderWithRedux(
      <RequestFormFieldBuilderContainer
        action={nonTypedActionInternal}
        id={"0"}
        tag={"WebActions"}
        typesMetadata={typesMetadata}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("Renders typed form int field", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "noRepeatedInt",
      types: testTypes
    })
    const { asFragment } = renderWithRedux(
      <RequestFormFieldBuilderContainer
        action={nonTypedActionInternal}
        id={"0"}
        tag={"WebActions"}
        typesMetadata={typesMetadata}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("Renders typed form nested int field", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "nestedNoRepeatedInt",
      types: testTypes
    })
    const { asFragment } = renderWithRedux(
      <RequestFormFieldBuilderContainer
        action={nonTypedActionInternal}
        id={"0"}
        tag={"WebActions"}
        typesMetadata={typesMetadata}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("Renders typed form repeated short field", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "repeatedShort",
      types: testTypes
    })
    const { asFragment } = renderWithRedux(
      <RequestFormFieldBuilderContainer
        action={nonTypedActionInternal}
        id={"0"}
        tag={"WebActions"}
        typesMetadata={typesMetadata}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("Renders typed form repeated nested int field", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "repeatedNestedNoRepeatedInt",
      types: testTypes
    })
    const { asFragment } = renderWithRedux(
      <RequestFormFieldBuilderContainer
        action={nonTypedActionInternal}
        id={"0"}
        tag={"WebActions"}
        typesMetadata={typesMetadata}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("Renders typed form repeated nested repeated short field", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "repeatedNestedRepeatedShort",
      types: testTypes
    })
    const { asFragment } = renderWithRedux(
      <RequestFormFieldBuilderContainer
        action={nonTypedActionInternal}
        id={"0"}
        tag={"WebActions"}
        typesMetadata={typesMetadata}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("Renders typed form multiple top level fields", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "multipleFlatFields",
      types: testTypes
    })
    const { asFragment } = renderWithRedux(
      <RequestFormFieldBuilderContainer
        action={nonTypedActionInternal}
        id={"0"}
        tag={"WebActions"}
        typesMetadata={typesMetadata}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
