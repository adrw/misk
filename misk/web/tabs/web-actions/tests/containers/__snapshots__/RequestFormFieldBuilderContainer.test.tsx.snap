// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestFormFieldBuilderContainer Doesn't fail on empty metadata 1`] = `
<DocumentFragment>
  <span />
</DocumentFragment>
`;

exports[`RequestFormFieldBuilderContainer Renders single body input 1`] = `
<DocumentFragment>
  <div
    class="bp3-control-group"
  >
    <span
      class="bp3-popover-wrapper"
    >
      <span
        class="bp3-popover-target"
      >
        <button
          class="bp3-button"
          tabindex="0"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            Body
          </span>
        </button>
      </span>
    </span>
    <span
      class="bp3-popover-wrapper"
    >
      <span
        class="bp3-popover-target"
      >
        <button
          class="bp3-button"
          tabindex="0"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-more"
            icon="more"
          >
            <svg
              data-icon="more"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                more
              </desc>
              <path
                d="M2 6.03a2 2 0 1 0 0 4 2 2 0 1 0 0-4zM14 6.03a2 2 0 1 0 0 4 2 2 0 1 0 0-4zM8 6.03a2 2 0 1 0 0 4 2 2 0 1 0 0-4z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
    </span>
    <div
      class="bp3-input-group"
    >
      <input
        class="bp3-input"
        placeholder="JSON"
        style="padding-right: 10px;"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`RequestFormFieldBuilderContainer Renders typed form int field 1`] = `
<DocumentFragment>
  <div
    class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
  >
    <div>
      <div
        class="bp3-control-group"
      >
        <span
          class="bp3-popover-wrapper"
        >
          <span
            class="bp3-popover-target"
          >
            <button
              class="bp3-button"
              tabindex="0"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Int Field
              </span>
            </button>
          </span>
        </span>
        <div
          class="bp3-input-group"
        >
          <input
            class="bp3-input"
            placeholder="Int"
            style="padding-right: 10px;"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RequestFormFieldBuilderContainer Renders typed form multiple top level fields 1`] = `
<DocumentFragment>
  <div
    class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
  >
    <div>
      <div
        class="bp3-control-group"
      >
        <span
          class="bp3-popover-wrapper"
        >
          <span
            class="bp3-popover-target"
          >
            <button
              class="bp3-button"
              tabindex="0"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Field 1
              </span>
            </button>
          </span>
        </span>
        <div
          class="bp3-input-group"
        >
          <input
            class="bp3-input"
            placeholder="Double"
            style="padding-right: 10px;"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div>
      <div
        class="bp3-control-group"
      >
        <span
          class="bp3-popover-wrapper"
        >
          <span
            class="bp3-popover-target"
          >
            <button
              class="bp3-button"
              tabindex="0"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Field 2
              </span>
            </button>
          </span>
        </span>
        <div
          class="bp3-input-group"
        >
          <input
            class="bp3-input"
            placeholder="Int"
            style="padding-right: 10px;"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div>
      <div
        class="bp3-control-group"
      >
        <span
          class="bp3-popover-wrapper"
        >
          <span
            class="bp3-popover-target"
          >
            <button
              class="bp3-button"
              tabindex="0"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Field 3
              </span>
            </button>
          </span>
        </span>
        <div
          class="bp3-input-group"
        >
          <input
            class="bp3-input"
            placeholder="Long"
            style="padding-right: 10px;"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div>
      <div
        class="bp3-control-group"
      >
        <span
          class="bp3-popover-wrapper"
        >
          <span
            class="bp3-popover-target"
          >
            <button
              class="bp3-button"
              tabindex="0"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Field 4
              </span>
            </button>
          </span>
        </span>
        <div
          class="bp3-input-group"
        >
          <input
            class="bp3-input"
            placeholder="Short"
            style="padding-right: 10px;"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RequestFormFieldBuilderContainer Renders typed form nested int field 1`] = `
<DocumentFragment>
  <div
    class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
  >
    <div>
      <div
        class="bp3-control-group"
      >
        <span
          class="bp3-popover-wrapper"
        >
          <span
            class="bp3-popover-target"
          >
            <button
              class="bp3-button"
              tabindex="0"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Nested Int Field
              </span>
            </button>
          </span>
        </span>
      </div>
      <div
        class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
      >
        <div>
          <div
            class="bp3-control-group"
          >
            <span
              class="bp3-popover-wrapper"
            >
              <span
                class="bp3-popover-target"
              >
                <button
                  class="bp3-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="bp3-button-text"
                  >
                    Int Field
                  </span>
                </button>
              </span>
            </span>
            <div
              class="bp3-input-group"
            >
              <input
                class="bp3-input"
                placeholder="Int"
                style="padding-right: 10px;"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RequestFormFieldBuilderContainer Renders typed form repeated double field 1`] = `
<DocumentFragment>
  <div
    class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
  >
    <div>
      <div>
        <div
          class="bp3-control-group"
        >
          <span
            class="bp3-popover-wrapper"
          >
            <span
              class="bp3-popover-target"
            >
              <button
                class="bp3-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-repeat"
                  icon="repeat"
                >
                  <svg
                    data-icon="repeat"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      repeat
                    </desc>
                    <path
                      d="M10 5c0 .55.45 1 1 1h4c.55 0 1-.45 1-1V1c0-.55-.45-1-1-1s-1 .45-1 1v1.74A7.95 7.95 0 0 0 8 0C3.58 0 0 3.58 0 8c0 4.06 3.02 7.4 6.94 7.92.02 0 .04.01.06.01.33.04.66.07 1 .07 4.42 0 8-3.58 8-8 0-.55-.45-1-1-1s-1 .45-1 1c0 3.31-2.69 6-6 6-.71 0-1.37-.15-2-.38v.01C3.67 12.81 2 10.61 2 8c0-3.31 2.69-6 6-6 1.77 0 3.36.78 4.46 2H11c-.55 0-1 .45-1 1z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                <span
                  class="bp3-button-text"
                >
                  Repeated Double Field
                </span>
              </button>
            </span>
          </span>
          <span
            class="bp3-popover-wrapper"
          >
            <span
              class="bp3-popover-target"
            >
              <button
                class="bp3-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-plus"
                  icon="plus"
                >
                  <svg
                    data-icon="plus"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      plus
                    </desc>
                    <path
                      d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
            </span>
          </span>
          <span />
          <div
            class="bp3-input-group"
          >
            <input
              class="bp3-input"
              placeholder="Double"
              style="padding-right: 10px;"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RequestFormFieldBuilderContainer Renders typed form repeated nested int field 1`] = `
<DocumentFragment>
  <div
    class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
  >
    <div>
      <div
        class="bp3-control-group"
      >
        <span />
      </div>
      <div
        class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
      >
        <div>
          <div
            class="bp3-control-group"
          >
            <span
              class="bp3-popover-wrapper"
            >
              <span
                class="bp3-popover-target"
              >
                <button
                  class="bp3-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="bp3-icon bp3-icon-repeat"
                    icon="repeat"
                  >
                    <svg
                      data-icon="repeat"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                    >
                      <desc>
                        repeat
                      </desc>
                      <path
                        d="M10 5c0 .55.45 1 1 1h4c.55 0 1-.45 1-1V1c0-.55-.45-1-1-1s-1 .45-1 1v1.74A7.95 7.95 0 0 0 8 0C3.58 0 0 3.58 0 8c0 4.06 3.02 7.4 6.94 7.92.02 0 .04.01.06.01.33.04.66.07 1 .07 4.42 0 8-3.58 8-8 0-.55-.45-1-1-1s-1 .45-1 1c0 3.31-2.69 6-6 6-.71 0-1.37-.15-2-.38v.01C3.67 12.81 2 10.61 2 8c0-3.31 2.69-6 6-6 1.77 0 3.36.78 4.46 2H11c-.55 0-1 .45-1 1z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                  <span
                    class="bp3-button-text"
                  >
                    Repeated Nested Int Field
                  </span>
                </button>
              </span>
            </span>
            <span
              class="bp3-popover-wrapper"
            >
              <span
                class="bp3-popover-target"
              >
                <button
                  class="bp3-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="bp3-icon bp3-icon-plus"
                    icon="plus"
                  >
                    <svg
                      data-icon="plus"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                    >
                      <desc>
                        plus
                      </desc>
                      <path
                        d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </button>
              </span>
            </span>
            <span />
          </div>
          <div
            class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
          >
            <div>
              <div
                class="bp3-control-group"
              >
                <span
                  class="bp3-popover-wrapper"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <button
                      class="bp3-button"
                      tabindex="0"
                      type="button"
                    >
                      <span
                        class="bp3-button-text"
                      >
                        Nested Int Field
                      </span>
                    </button>
                  </span>
                </span>
              </div>
              <div
                class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
              >
                <div>
                  <div
                    class="bp3-control-group"
                  >
                    <span
                      class="bp3-popover-wrapper"
                    >
                      <span
                        class="bp3-popover-target"
                      >
                        <button
                          class="bp3-button"
                          tabindex="0"
                          type="button"
                        >
                          <span
                            class="bp3-button-text"
                          >
                            Int Field
                          </span>
                        </button>
                      </span>
                    </span>
                    <div
                      class="bp3-input-group"
                    >
                      <input
                        class="bp3-input"
                        placeholder="Int"
                        style="padding-right: 10px;"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RequestFormFieldBuilderContainer Renders typed form repeated nested repeated double field 1`] = `
<DocumentFragment>
  <div
    class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
  >
    <div>
      <div
        class="bp3-control-group"
      >
        <span />
      </div>
      <div
        class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
      >
        <div>
          <div
            class="bp3-control-group"
          >
            <span
              class="bp3-popover-wrapper"
            >
              <span
                class="bp3-popover-target"
              >
                <button
                  class="bp3-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="bp3-icon bp3-icon-repeat"
                    icon="repeat"
                  >
                    <svg
                      data-icon="repeat"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                    >
                      <desc>
                        repeat
                      </desc>
                      <path
                        d="M10 5c0 .55.45 1 1 1h4c.55 0 1-.45 1-1V1c0-.55-.45-1-1-1s-1 .45-1 1v1.74A7.95 7.95 0 0 0 8 0C3.58 0 0 3.58 0 8c0 4.06 3.02 7.4 6.94 7.92.02 0 .04.01.06.01.33.04.66.07 1 .07 4.42 0 8-3.58 8-8 0-.55-.45-1-1-1s-1 .45-1 1c0 3.31-2.69 6-6 6-.71 0-1.37-.15-2-.38v.01C3.67 12.81 2 10.61 2 8c0-3.31 2.69-6 6-6 1.77 0 3.36.78 4.46 2H11c-.55 0-1 .45-1 1z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                  <span
                    class="bp3-button-text"
                  >
                    Nested Repeated Double Field
                  </span>
                </button>
              </span>
            </span>
            <span
              class="bp3-popover-wrapper"
            >
              <span
                class="bp3-popover-target"
              >
                <button
                  class="bp3-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="bp3-icon bp3-icon-plus"
                    icon="plus"
                  >
                    <svg
                      data-icon="plus"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                    >
                      <desc>
                        plus
                      </desc>
                      <path
                        d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </button>
              </span>
            </span>
            <span />
          </div>
          <div
            class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
          >
            <div>
              <div
                class="bp3-control-group"
              >
                <span
                  class="bp3-popover-wrapper"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <button
                      class="bp3-button"
                      tabindex="0"
                      type="button"
                    >
                      <span
                        class="bp3-button-text"
                      >
                        Nested Double Field
                      </span>
                    </button>
                  </span>
                </span>
              </div>
              <div
                class="bp3-card bp3-elevation-0 sc-bZQynM cLWGOp"
              >
                <div>
                  <div>
                    <div
                      class="bp3-control-group"
                    >
                      <span
                        class="bp3-popover-wrapper"
                      >
                        <span
                          class="bp3-popover-target"
                        >
                          <button
                            class="bp3-button"
                            tabindex="0"
                            type="button"
                          >
                            <span
                              class="bp3-icon bp3-icon-repeat"
                              icon="repeat"
                            >
                              <svg
                                data-icon="repeat"
                                height="16"
                                viewBox="0 0 16 16"
                                width="16"
                              >
                                <desc>
                                  repeat
                                </desc>
                                <path
                                  d="M10 5c0 .55.45 1 1 1h4c.55 0 1-.45 1-1V1c0-.55-.45-1-1-1s-1 .45-1 1v1.74A7.95 7.95 0 0 0 8 0C3.58 0 0 3.58 0 8c0 4.06 3.02 7.4 6.94 7.92.02 0 .04.01.06.01.33.04.66.07 1 .07 4.42 0 8-3.58 8-8 0-.55-.45-1-1-1s-1 .45-1 1c0 3.31-2.69 6-6 6-.71 0-1.37-.15-2-.38v.01C3.67 12.81 2 10.61 2 8c0-3.31 2.69-6 6-6 1.77 0 3.36.78 4.46 2H11c-.55 0-1 .45-1 1z"
                                  fill-rule="evenodd"
                                />
                              </svg>
                            </span>
                            <span
                              class="bp3-button-text"
                            >
                              Repeated Double Field
                            </span>
                          </button>
                        </span>
                      </span>
                      <span
                        class="bp3-popover-wrapper"
                      >
                        <span
                          class="bp3-popover-target"
                        >
                          <button
                            class="bp3-button"
                            tabindex="0"
                            type="button"
                          >
                            <span
                              class="bp3-icon bp3-icon-plus"
                              icon="plus"
                            >
                              <svg
                                data-icon="plus"
                                height="16"
                                viewBox="0 0 16 16"
                                width="16"
                              >
                                <desc>
                                  plus
                                </desc>
                                <path
                                  d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                                  fill-rule="evenodd"
                                />
                              </svg>
                            </span>
                          </button>
                        </span>
                      </span>
                      <span />
                      <div
                        class="bp3-input-group"
                      >
                        <input
                          class="bp3-input"
                          placeholder="Double"
                          style="padding-right: 10px;"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
